src/main/java/io/kestra/plugin/serdes/avro/infer/InferAvroSchema.java

```diff
@@ -0,0 +1,257 @@
+                .reduce(
+                    InferAvroSchema::mergeTypes
+                )
+                .map(Field::schema);
+        } catch (IOException e) {
+            throw new RuntimeException("could not parse Ion input stream, err: " + e.getMessage(), e);
+        }
+        try {
+            output.write(inferedSchema.block().toString().getBytes(StandardCharsets.UTF_8));
+        } catch (IOException e) {
+            throw new RuntimeException("could not write Avro schema in output stream, err: " + e.getMessage(), e);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private Field inferField(String fieldFullPath, String fieldName, Object node) {
+        Field inferredField = null;
+        if (node instanceof Map) {
+            var map = (Map<String, Object>) node;
+            var inferredFields = new ArrayList<Field>();
+            for (Map.Entry<String, Object> field : map.entrySet()) {
+                inferredFields.add(
+                        inferField(
+                                fieldFullPath + "_" + fieldName + "_" + field.getKey(),
+                                field.getKey(),
+                                field.getValue()
+                        )
+                );
+            }
+
+            var recordSchema = Schema.createRecord(
+                fieldName,
+                null,
+                "io.kestra.plugin.serdes.avro",
+                false,
+                inferredFields
+            );
+            if ("root".equals(fieldName)) {
+                inferredField = new Field(
+                    fieldName,
+                    recordSchema
+                );
+            } else {
+                inferredField = new Field(
+                    fieldName,
+                    Schema.createUnion(
+                        recordSchema,
+                        Schema.create(Schema.Type.NULL)
+                    )
+                );
+            }
+        } else if (node instanceof List) {
+            var list = (List<Object>) node;
+            if (!list.isEmpty()) {
+                Field inferredType = null;
+                if (deepSearch) {
+                    for (Object item : list) {
+                        inferredType = inferField(fieldFullPath + "_" + fieldName + "_items", fieldName + "_items", item);
+                    }
+                } else {
+                    inferredType = inferField(fieldFullPath + "_" + fieldName + "_items", fieldName + "_items", list.get(0));
+                }
+                if ("root".equals(fieldName)) {
+                    inferredField = new Field(
+                        fieldName,
+                        Schema.createArray(inferredType.schema()),
+                        "",
+                        Collections.emptyList()
+                    );
+                } else {
+                    inferredField = new Field(
+                        fieldName,
+                        Schema.createUnion(
```

Comment:
When creating a UNION type, `null` should always be the first element. In Avro, the default value for a field must be of the first type of the union

Reply:
I know 😭, but when doing so our Ion parser has issues and for example convert String to Byte

Reply:
should be fixed